File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
  public:
    int furthestBuilding(vector < int > & heights, int bricks, int ladders) {
      int n = heights.size();
      int totalBricksUsed = 0;
      priority_queue < int > bricksUsed;
      int reach = 0;
      for (; reach < n - 1; reach++) {
        // If current height is higher, we don't have to use
        // bricks or ladder.
        if (heights[reach] >= heights[reach + 1]) {
          continue;
        }

        // We have to use either brick or ladder.
        int diff = heights[reach + 1] - heights[reach];

        if (totalBricksUsed + diff <= bricks) {
          // Prefer bricks because we can later change from 
          // bricks to ladder as we will see below.
          totalBricksUsed += diff;
          bricksUsed.push(diff);
        } else if (ladders > 0) {
          // If even bricks are insufficient, check ladders.
          if (!bricksUsed.empty() && bricksUsed.top() > diff) {
            totalBricksUsed = totalBricksUsed - bricksUsed.top() + diff;
            bricksUsed.pop();
            bricksUsed.push(diff);
          }
          ladders--;
        } else {
          // Unfortunately, it's not possible to go ahead any further.
          break;
        }
      }
      return reach; // This is the max we can go ahead.
    }
};

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        int n = heights.length;
        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for (int i = 0; i < n - 1; i++) {
            int diff = heights[i + 1] - heights[i];
            if (diff > 0) { // If the next building is taller than the current one
                pq.add(diff); // Add the height difference to the priority queue
                if (pq.size() > ladders) { // If there are more height differences than available ladders
                    bricks -= pq.poll(); // Use a ladder by removing the smallest height difference from the priority queue
                    if (bricks < 0) { // If not enough bricks to bridge the remaining gaps
                        return i; // Return the index of the current building (furthest reachable building)
                    }
                }
            }
        }

        return n - 1; // If all buildings can be reached, return the index of the last building
    }
}